apollo-router/src/plugins/cache/invalidation.rs

```diff
@@ -0,0 +1,151 @@
+        if let Some(storage) = storage {
+            let h = handle.clone();
+
+            tokio::task::spawn(async move { start(storage, h.into_stream()).await });
+        }
+        Ok(Self { enabled, handle })
+    }
+
+    pub(crate) async fn invalidate(
+        &mut self,
+        requests: Vec<InvalidationRequest>,
+    ) -> Result<(), BoxError> {
+        if self.enabled {
+            let mut sink = self.handle.clone().into_sink();
+            sink.send(requests).await.map_err(|e| e.message)?;
+        }
+
+        Ok(())
+    }
+}
+
+async fn start(
+    storage: RedisCacheStorage,
+    mut handle: HandleStream<InvalidationTopic, Vec<InvalidationRequest>>,
+) {
+    while let Some(requests) = handle.next().await {
+        handle_request_batch(&storage, requests)
+            .instrument(tracing::info_span!("cache.invalidation.batch"))
+            .await
+    }
+}
+
+async fn handle_request_batch(storage: &RedisCacheStorage, requests: Vec<InvalidationRequest>) {
+    for request in requests {
+        let start = Instant::now();
+        handle_request(storage, &request)
+            .instrument(tracing::info_span!("cache.invalidation.request"))
+            .await;
+        f64_histogram!(
+            "apollo.router.cache.invalidation.duration",
+            "Duration of the invalidation event execution.",
+            start.elapsed().as_secs_f64()
+        );
+    }
+}
+
+async fn handle_request(storage: &RedisCacheStorage, request: &InvalidationRequest) {
+    let key_prefix = request.key_prefix();
+    tracing::debug!(
+        "got invalidation request: {request:?}, will scan for: {}",
+        key_prefix
+    );
+
+    // FIXME: configurable batch size
+    let mut stream = storage.scan(key_prefix.clone(), Some(10));
+    let mut count = 0u64;
+
+    while let Some(res) = stream.next().await {
+        match res {
+            Err(e) => {
+                tracing::error!(
+                    pattern = key_prefix,
+                    error = %e,
+                    message = "error scanning for key",
+                );
+                break;
+            }
+            Ok(scan_res) => {
+                if let Some(keys) = scan_res.results() {
+                    let keys = keys
+                        .iter()
+                        .filter_map(|k| k.as_str())
+                        .map(|k| RedisKey(k.to_string()))
+                        .collect::<Vec<_>>();
+                    if !keys.is_empty() {
+                        tracing::debug!("deleting keys: {keys:?}");
+                        count += keys.len() as u64;
+                        storage.delete(keys).await;
+                    }
+                }
+            }
+        }
+    }
+
+    u64_histogram!(
+        "apollo.router.cache.invalidation.keys",
+        "Number of invalidated keys.",
+        count
+    );
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(tag = "kind")]
+pub(crate) enum InvalidationRequest {
```

Comment:
Do we really want to have the kind field capitalized ?

Reply:
do I do that with `#[serde(rename_all = "lowercase")]` over the whole enum?

Reply:
done in a189c67bd